### ISSUE TITLE
FifoDiskQueue Size is Getting Bigger

### ISSUE THREAD

**Issue by DavidDr90** [scrapy/queuelib#57]
I'm trying to implement a fixed size queue using FifoDiskQueue. 
My code is:

`self._images_data = FifoDiskQueue(r"c:\temp")`
`...`
`# this code is inside a loop`
`# implement fixed size queue`
`if len(self._images_data) >= 400:`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`self._images_data.pop()`
`self._images_data.push(frame.tobytes())  # add the new frame to the main data base`
`...`
`# after the while ends running:`
`self._images_data.close()`

I'm seeing two issues:
1. While the main loop is running the queue file `q00000` is always 0 KB
2. My memory is decreasing meaning the queue is keep getting bigger and bigger although I'm "pop" out any old element before adding new one 

**Can some help me figure out why the size keep getting bigger or point me to the right way to implement a fixed size queue?**

**Comment by nisseknudsen** [comment]
tl;dr: `FifoDiskQueue(..., chunksize=1)` avoid an ever incrementing file, at the cost of having a ton more files on disk.

-----

I know it's 3 years late, but maybe it helps the next person:

Ran into the same issue today: the queue file on disk (seemingly) does not reduce its size after items are being popped. I looked at the function calls and it seems like a design choice.

A `FifoDiskQueue` object has a constructor parameter called `chunk_size: int = 100_000`. This parameter is a basically the amount of files that need to be popped until a new queue on disk is being created. So for my use case, this was way too high (I only tested with low hundreds of files), so it looked like the queue grows ever so larger.

Here's the script I used and what the queue on disk looked like (the .mp4 file is ~34mbyte):

```
import time

from queuelib import FifoDiskQueue

import tempfile

tmp_dir = tempfile.mkdtemp()
queue = FifoDiskQueue(tmp_dir, chunksize=10)

# print own process ID
print(f"ls -l {tmp_dir}")

with open("20250307_225512_461_+0000.mp4", "rb") as fp:
    data = fp.read()

for j in range(2):
    for i in range(10):
        queue.push(data)

    print("Queue size:", len(queue))
    time.sleep(10)

    while len(queue) > 0:
        item = queue.pop()

    print("Queue size:", len(queue))
    time.sleep(10)

queue.close()  # <-- only here the queue file gets removed from disk
```

with the following queue file on disk:

1. After pushing 10 files:
```
-rw-r--r--  1 nisseknudsen  staff  322858080 Mar 12 20:50 q00000
```
2. After popping 10 files:
```
-rw-r--r--  1 nisseknudsen  staff  645716160 Mar 12 20:50 q00000 <-- I was expecting this to be close to 0 bytes
```
3. After pushing 10 files:
```
-rw-r--r--  1 nisseknudsen  staff  645716160 Mar 12 20:50 q00000 <-- I was expecting this to be now 322858080 bytes
```


Now, with that knowlege, I then set the `chunk_size=10` and repeated the test:

```
# same script as before, except:
...
queue = FifoDiskQueue(tmp_dir, chunksize=10)
...
```

and now it looked like this on disk:

1. After pushing 10 files:
```
-rw-r--r--  1 nisseknudsen  staff  322858080 Mar 12 20:58 q00000
-rw-r--r--  1 nisseknudsen  staff          0 Mar 12 20:58 q00001  <--  the `.push` method already prepared the next chunk because we hit the limit exactly with 10 files
```
2. After popping 10 files:
```
-rw-r--r--  1 nisseknudsen  staff  0 Mar 12 20:58 q00001 <-- you can see that the first queue file has been completely deleted, only the next queue file waiting for us
```
3. After pushing 10 files:
```
-rw-r--r--  1 nisseknudsen  staff  322858080 Mar 12 20:58 q00001 <-- Correctly filled
-rw-r--r--  1 nisseknudsen  staff          0 Mar 12 20:58 q00002 <-- Next queue already in the making
```


So how does this help?
I'm not sure - if the file size per item on my end was constant, I could probably find an optimal `chunk_size`.
**If you don't care about having multiple files on the disk, you could always go with `chunk_size=1` and you probably have what you were looking for**. I did this as another test, see results below:

1. After pushing 10 files:
```
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00000
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00001
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00002
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00003
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00004
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00005
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00006
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00007
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00008
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00009
-rw-r--r--  1 nisseknudsen  staff         0 Mar 12 21:03 q00010
```
2. After popping 10 files:
```
-rw-r--r--  1 nisseknudsen  staff  0 Mar 12 21:03 q00010
```
3. After pushing 10 files
```
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00010
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00011
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00012
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00013
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00014
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00015
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00016
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00017
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00018
-rw-r--r--  1 nisseknudsen  staff  32285808 Mar 12 21:03 q00019
-rw-r--r--  1 nisseknudsen  staff         0 Mar 12 21:03 q00020
```
